qiskit_experiments/test/mock_iq_backend.py

```diff
@@ -193,43 +306,146 @@ def _default_options(self):
             meas_return="single",
         )
 
-    def _draw_iq_shots(self, prob, shots, phase: float = 0.0) -> List[List[List[float]]]:
-        """Produce an IQ shot."""
-
-        rand_i = self._rng.normal(0, self._iq_cluster_width, size=shots)
-        rand_q = self._rng.normal(0, self._iq_cluster_width, size=shots)
+    def set_calculation_parameters(self, calculation_parameters: List[Dict[str, Any]]):
+        """
+        A function to set calculation parameters for the compute_probabilities function.
+        Args:
+            calculation_parameters(Dict): A dictionary with parameters that will replace the existing
+            parameters of the backend.
+        """
+        self._calculation_parameters = calculation_parameters
 
-        memory = []
-        for idx, state in enumerate(self._rng.binomial(1, prob, size=shots)):
+    @staticmethod
+    def _verify_parameters(num_qubits: int, prob_dict: Dict[str, float]):
+        if num_qubits < 1:
+            raise ValueError(f"The number of qubits {num_qubits} is smaller than 1.")
+
+        if not np.allclose(1, sum(prob_dict.values())):
+            raise ValueError("The probabilities given don't sum up to 1.")
+        for key in prob_dict.keys():
+            if num_qubits is not len(key):
+                raise ValueError(
+                    "The number of qubits and the length of the output string don't match."
+                )
+
+    def _get_normal_samples_for_shot(self, num_qubits: int):
+        """
+        Produce a list in the size of num_qubits. Each entry value is produced from normal distribution
+        with expected value of '0' and standard deviation of self._iq_cluster_width.
+        Args:
+            num_qubits(int): The amount of qubits in the circuit.
 
-            if state > 0.5:
-                point_i = self._iq_cluster_centers[0] + rand_i[idx]
-                point_q = self._iq_cluster_centers[1] + rand_q[idx]
-            else:
-                point_i = self._iq_cluster_centers[2] + rand_i[idx]
-                point_q = self._iq_cluster_centers[3] + rand_q[idx]
+        Returns:
+            List[float]: A list with values that were produced from normal distribution.
+        """
+        widths = self._iq_cluster_width
+        samples = [self._rng.normal(0, widths[qubit], size=1) for qubit in range(num_qubits)]
+        # we squeeze the second dimension because samples is List[qubit_number][0][0\1] = I\Q
+        # and we want to change it to be List[qubit_number][0\1]
+        return np.squeeze(np.array(samples), axis=1)
+
+    def _probability_dict_to_probability_array(
+        self, prob_dict: Dict[str, float], num_qubits: int
+    ) -> List[float]:
+        prob_list = []
+        for num in range(2**num_qubits):
+            num_binary_str = str(format(num, "b").zfill(num_qubits))
+            prob_list.append(prob_dict.get(num_binary_str, 0))
```

Comment:
I'm not sure how it is in Python. In other languages it is recommended to initialize large arrays instead of incrementally appending to them. This prevent a series of reallocations.
```suggestion
        prob_list = [0] * 2**num_qubits
        for output_str in prob_dict:
            index = < compute the index of the output string >
            prob_list[index] = prob_dict[output_str]
```

Reply:
Good idea. changed to the following code:
```python
    def _probability_dict_to_probability_array(
        self, prob_dict: Dict[str, float], num_qubits: int
    ) -> List[float]:
        prob_list = [0] * (2**num_qubits)
        for output_str, probability in prob_dict.items():
            index = int(output_str, 2)
            prob_list[index] = probability
        return prob_list
```